File: csl-rfid-android-sdk/src/main/java/com/csl/rfidsdk/models/RfidGeigerStats.java
package com.csl.rfidsdk.models;

/**
 * Statistics for RFID Geiger search operations
 */
public class RfidGeigerStats {
    private final double currentRssi;
    private final double peakRssi;
    private final int readCount;
    private final long duration;
    private final double readRate;

    public RfidGeigerStats(double currentRssi, double peakRssi, int readCount, long duration, double readRate) {
        this.currentRssi = currentRssi;
        this.peakRssi = peakRssi;
        this.readCount = readCount;
        this.duration = duration;
        this.readRate = readRate;
    }

    public double getCurrentRssi() {
        return currentRssi;
    }

    public double getPeakRssi() {
        return peakRssi;
    }

    public int getReadCount() {
        return readCount;
    }

    public long getDuration() {
        return duration;
    }

    public double getReadRate() {
        return readRate;
    }

    /**
     * Get proximity level (0.0 - 1.0)
     * Based on RSSI threshold mapping
     * -90 dBm = 0.0 (far)
     * -10 dBm = 1.0 (very close)
     */
    public double getProximityLevel() {
        double min = -90.0;
        double max = -10.0;
        double proximity = (currentRssi - min) / (max - min);
        return Math.max(0.0, Math.min(1.0, proximity));
    }

    @Override
    public String toString() {
        return "RfidGeigerStats{" +
                "currentRssi=" + currentRssi +
                ", peakRssi=" + peakRssi +
                ", readCount=" + readCount +
                ", duration=" + duration +
                ", readRate=" + readRate +
                ", proximityLevel=" + getProximityLevel() +
                '}';
    }

    /**
     * Builder for RfidGeigerStats
     */
    public static class Builder {
        private String targetEpc;
        private double currentRssi;
        private double peakRssi;
        private int readCount;
        private int proximity;
        private long elapsedTimeMs;

        public Builder targetEpc(String targetEpc) {
            this.targetEpc = targetEpc;
            return this;
        }

        public Builder currentRssi(double currentRssi) {
            this.currentRssi = currentRssi;
            return this;
        }

        public Builder peakRssi(double peakRssi) {
            this.peakRssi = peakRssi;
            return this;
        }

        public Builder readCount(int readCount) {
            this.readCount = readCount;
            return this;
        }

        public Builder proximity(int proximity) {
            this.proximity = proximity;
            return this;
        }

        public Builder elapsedTimeMs(long elapsedTimeMs) {
            this.elapsedTimeMs = elapsedTimeMs;
            return this;
        }

        public RfidGeigerStats build() {
            double readRate = 0.0;
            if (elapsedTimeMs > 0) {
                readRate = (readCount * 1000.0) / elapsedTimeMs;
            }
            return new RfidGeigerStats(currentRssi, peakRssi, readCount, elapsedTimeMs, readRate);
        }
    }
}

File: cs710flutterapp/lib/models/rfid_geiger_stats.dart
/// Statistics for Geiger search (tag locating) operations
class RfidGeigerStats {
  final String targetEpc;
  final double currentRssi;
  final double peakRssi;
  final int readCount;
  final int proximity; // 0-100%
  final int elapsedTimeMs;

  const RfidGeigerStats({
    required this.targetEpc,
    required this.currentRssi,
    required this.peakRssi,
    required this.readCount,
    required this.proximity,
    required this.elapsedTimeMs,
  });

  /// Create RfidGeigerStats from platform channel map
  factory RfidGeigerStats.fromMap(Map<String, dynamic> map) {
    return RfidGeigerStats(
      targetEpc: map['targetEpc'] as String,
      currentRssi: (map['currentRssi'] as num).toDouble(),
      peakRssi: (map['peakRssi'] as num).toDouble(),
      readCount: map['readCount'] as int,
      proximity: map['proximity'] as int,
      elapsedTimeMs: map['elapsedTimeMs'] as int,
    );
  }

  /// Convert to map for platform channel
  Map<String, dynamic> toMap() {
    return {
      'targetEpc': targetEpc,
      'currentRssi': currentRssi,
      'peakRssi': peakRssi,
      'readCount': readCount,
      'proximity': proximity,
      'elapsedTimeMs': elapsedTimeMs,
    };
  }

  /// Get elapsed time in seconds
  int get elapsedSeconds => (elapsedTimeMs / 1000).floor();

  @override
  String toString() =>
      'RfidGeigerStats(epc: $targetEpc, rssi: $currentRssi, proximity: $proximity%)';
}
